### PR TITLE
AzureRmVMDiagnostics command change [yantang]

### PR DESCRIPTION
**Pull request by yantang-msft**
Hi, this code is changed by Microsoft FTE from azure tools team. Alias: yantang
The change aimed at making AzureRmVMDiagnostics related command more reasonable to use, also fixed some bugs. Detailed changes are listed below.

Set-AzureRMVMDiagnosticsExtension
    a) Make StorageContext optional, and make -StorageAccountName -StorageAccountKey as optional parameter
        a. Can parse account name from config file if not specified
        b. Can get the account key if not specified and account is under current subs
        c. Handle storage endpoint address.
    b) Fix bug so that user can use a storage account in a different subscription other than the current one.
    c) Make -Name an optional parameter, default to "Microsoft.Insights.VMDiagnosticsSettings"
    d) Make -Location optional, default to using the VM's location.
    e) -Version Optional parameter, default is 1.5
    f) -AutoUpgradeMinorVersion optional, default is true

Get-AzureRMVMDiagnosticsExtension & Remove-AzureRMVMDiagnosticsExtension
    a) Make -Name optional parameter

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@yantang-msft__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

**Comment by yantang-msft** [comment]
For testing, Get and Remove is fairly easy, just specify the ResourceGroupName and VMName name. I've tested locally for the case with/without diagnostics extension.
For Set, ResourceGroupName, VMName, DiagnosticsConfigurationFilePath is required. Other parameters should be automatically resolved in most cases. The only special parameter is StorageAccountKey. If the StorageAccountName is not within current subscription, then user must specify the key as well since we can't resolve it.
I've tested the case for using storage within same resource group, within current subscription but different resource group and storage in another subscription. I can see the diagnostics logs sent to the storage tables.

There is also a known issue, if there is already an diagnostics extension, user must first Remove the extension then Set.

Let me know if you have any other questions (alias: yantang).

**Comment by markcowl** [comment]
@azuresdkci add to whitelist

File: src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+using Microsoft.WindowsAzure.Commands.Utilities.Common;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Auth;
+
+namespace Microsoft.Azure.Commands.Compute
+{
+    [Cmdlet(
+        VerbsCommon.Set,
+        ProfileNouns.VirtualMachineDiagnosticsExtension)]
+    public class SetAzureVMDiagnosticsExtensionCommand : VirtualMachineExtensionBaseCmdlet
```
Comment (hovsepm) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs]: please change the class name to reflect cmdlet name change.

**Comment by hovsepm** [comment]
@yantang-msft please change all class names in your PR to reflect cmdlet name (see my comment in SetAzureRmVMDiagnosticsExtension.cs).

**Comment by hovsepm** [comment]
Also add tests to verify changed functionality.

**Comment by markcowl** [comment]
@yantang-msft Please address review feedback.  Note that code complete is EOD today, so this PR is at risk of slipping until the January release

**Comment by yantang-msft** [comment]
@hovsepm @markcowl Comments addressed, with "adding test case" pending.
We've manually tested E2E scenarios within our team. Will send another pull request to add test cases, hopefully this week.

File: src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs (hunk 1 above)
Comment (yantang-msft) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs]: Resolved

**Pull request merged**